magma/array.py

```diff
@@ -222,6 +222,21 @@ def __getitem__(self, key):
 
             return self.ts[key]
 
+    def __setitem__(self, key, val):
+        error = False
+        old = self[key]
+        if isinstance(old, Array):
+            if len(old) != len(val):
+                error = True
+            elif any(old[i] is not val[i] for i in range(len(old))):
+                error = True
+        elif old is not val:
+            error = True
+
+        if error:
+            raise ValueError('May not mutate array')
```

Comment:
Since this would likely arise in user facing code, let's make this a logged error: `_logger.Error("Directly calling setitem on Array disallowed")`

Reply:
/I am fine with this, however it would change current behavior. `Array[4, Bit]()[0] = Bit(0)` is a `TypeError` currently.

I also don't know how to use the logger can you point me at example code?

Reply:
Why would this depend on whether we `raise` or log?

Usage of logger: https://github.com/phanrahan/magma/blob/77689a30e595e343c8ae4030d6cb20fd0206b834/magma/port.py#L39-L40. We define a `_logger` at the top of this file, and it has methods `error`/`log`/`warning`

Reply:
@rsetaluri I was just saying it changes code that previously raised into code that logs.  Which doesn't matter to me, but it does change behavior.